include/nbl/system/IAsyncQueueDispatcher.h

```diff
@@ -19,10 +19,20 @@ namespace impl
             // wait on this for result to be ready
             std::condition_variable cvar;
             bool ready = false;
+
+            void lock() {}
+            void unlock() {}
         };
     };
 }
 
+/**
+* Provided RequestType may optionally define 2 member functions:
+* void lock();
+* void unlock();
+* lock() will be called just before processing the request, and unlock() will be called just after processing the request.
+* Those are to enable safe external write access to the request struct for user-defined purposes.
+*/
 template <typename CRTP, typename RequestType, uint32_t BufferSize = 256u, typename InternalStateType = void>
 class IAsyncQueueDispatcher : public IThreadHandler<CRTP, InternalStateType>, public impl::IAsyncQueueDispatcherBase
```

Comment:
I'd add some failure handling function when you're about to overflow in `IAsyncQueueDispatcher::request`, cause a nasty overflow with no fallbacks is evil.

Also don't `cb_begin` and `cb_end` need to be atomic?

Reply:
you should add some static asserts that the `RequestType` is non-copyable and non-movable

Reply:
> 
> 
> I'd add some failure handling function when you're about to overflow in `IAsyncQueueDispatcher::request`, cause a nasty overflow with no fallbacks is evil.
> 
> Also don't `cb_begin` and `cb_end` need to be atomic?

you could push and pop from the circular buffer without a condvar

Reply:
> I'd add some failure handling function when you're about to overflow in IAsyncQueueDispatcher::request, cause a nasty overflow with no fallbacks is evil.

there should just be another cvar for overflow cases (thread's waiting until there's a free space in circular buffer), but i didnt implement it yet. This requires FIFO behaviour though, but it's not that hard to do either

> Also don't cb_begin and cb_end need to be atomic?

no, only ever touched under locked mutex

Reply:
> no, only ever touched under locked mutex

plus just having it atomic is not enough because of need to wrap around

Reply:
> 
> 
> > no, only ever touched under locked mutex
> 
> plus just having it atomic is not enough because of need to wrap around

wrapping around has nothing to do with the atomicity.

because wrapping around is what you do to the already incremented and fetched value

Reply:
> no, only ever touched under locked mutex

If they were atomic you'd never need a lock, it would be a lockless Circular Buffer

Reply:
> there should just be another cvar for overflow cases (thread's waiting until there's a free space in circular buffer), but i didnt implement it yet. This requires FIFO behaviour though, but it's not that hard to do either

this one is kinda hard (FIFO wakeups), although not sure if threads should expect that their submissions be FIFO (not preempted). Its probably a problem that we should only solve if we ever suffer from it.

We'd probably do basic part by
```cpp
submit()
{
// The "virtual" element can always be reserved
auto virtualIx = cb_end++;

static_assert(BufferSize>1u,"Because executor only notifies about beginning to process a request, and nothing is known about when its finished aside from that it will be finished when the next request begins to be procesed, this class always \"wastes\" 1 element of request storage.");
auto safeBegin = virtualIx<BufferSize ? 0u:(virtualIx-BufferSize+1u); // 1u because the memory needs to be alive during the processing of a request. This works because there's only one executor
for (decltype(cb_begin)::underlying_t oldBegin; (oldBegin=cb_begin.load())<safeBegin; )
{
#if __cplusplus >= 202002L
    cb_begin.wait(oldBegin);
#else
    std::this_thread::yield();
#endif
}
// now we can be sure that we wont overwrite the memory used by any pending request

auto placeIx = wrapAround(virtualIx);
requests[placeIx] = /*place/construct the request*/;
}

execute()
{
auto reqIx = wrapAround(cb_begin++);
#if __cplusplus >= 202002L
cb_begin.notify_one();
#endif

// holding onto a reference without copying works only because there's one executor thread (single consumer)
auto& request = requests[reqIx];
// do work on request
}
```
No

Reply:
> its just that name "Queue" suggests execution in specific order

true, for Video queues it matters.

Reply:
> oh, ok, if we dont want to keep cb_begin and cb_end wrapped around then its no problem. Perfectly right in case of circular buffer size being PoT

Yeah PoT is nice for multiple reasons.
Yeah PoT is nice for multiple reasons ;)

Reply:
kinda resolved, but no word from @Crisspl whether it works.

Reply:
yep, it's working